parlai/core/torch_generator_agent.py

```diff
@@ -0,0 +1,1022 @@
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+
+"""
+**BETA**: This module is in Beta. Feedback is most welcome, and the API is highly
+likely to change.
+
+
+Generic Pytorch-based Generator agent. Implements quite a bit of boilerplate,
+including Beam search.
+
+Contains the following utilities:
+
+* TorchGeneratorAgent class, which serves as a useful parent for generative torch
+  agents.
+* Beam class which provides some generic beam functionality for classes to use
+
+TODO: Docs.
+"""
+
+import os
+import math
+import tempfile
+from collections import defaultdict, Counter, namedtuple
+from operator import attrgetter
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from parlai.core.torch_agent import TorchAgent, Output
+from parlai.core.utils import NEAR_INF, padded_tensor, round_sigfigs
+from parlai.core.thread_utils import SharedTable
+
+# warn the user if they run with --skip-generation, but only once
+_WARNING_SEEN = False
+
+
+def _pad_to_length(tensor, length, dim=0, pad=0):
+    """Pad tensor to a specific length.
+
+    :param tensor: vector to pad
+    :param length: new length
+    :param dim: (default 0) dimension to pad
+
+    :returns: padded tensor if the tensor is shorter than length
+    """
+    if tensor.size(dim) < length:
+        return torch.cat(
+            [tensor, tensor.new(*tensor.size()[:dim],
+                                length - tensor.size(dim),
+                                *tensor.size()[dim + 1:]).fill_(pad)],
+            dim=dim)
+    else:
+        return tensor
+
+
+class TorchGeneratorModel(nn.Module):
+    """
+    This Interface expects you to implement model with the following reqs:
+
+    :field model.encoder: takes input returns tuple (enc_out, enc_hidden, attn_mask)
+    :field model.decoder: takes decoder params and returns decoder outputs after attn
+    :field model.output: takes decoder outputs and returns distr over dictionary
+    """
+    def __init__(
+        self,
+        padding_idx=0,
+        start_idx=1,
+        unknown_idx=3,
+        input_dropout=0,
+        longest_label=1,
+    ):
+        super().__init__()
+        self.NULL_IDX = padding_idx
+        self.register_buffer('START', torch.LongTensor([start_idx]))
+        self.longest_label = longest_label
+
+    def _starts(self, bsz):
+        """Return bsz start tokens."""
+        return self.START.detach().expand(bsz, 1)
+
+    def greedy_decode(self, bsz, encoder_states, maxlen):
+        """Greedy search"""
+        xs = self._starts(bsz)
+        incr_state = None
+        for _ in range(maxlen):
+            scores, incr_state = self.decoder(xs, encoder_states, incr_state)
+            _, preds = self.output(scores).max(dim=-1)
+            xs = torch.cat([xs, preds], dim=1)
```

Comment:
I think the API for greedy decoding makes more sense to just pass in the current token rather than all the tokens so far (then we also won't have to cat at every time step), and then the seq2seq module decode doesn't have to `xs = xs[: -1:]`

Reply:
So the reason for this is that not all models will support incremental decoding -- we should use it if it's available, but fall back O(n^2) generation if we have to. Let's talk about clean ways to distinguish?